### DISCUSSION TITLE
Dynamically changing Laravel Echo channel listeners doesn't work

### DISCUSSION THREAD

**Discussion by waterbuckit** [livewire/livewire#3115] [7. Bugs (archived)]
### Description
If for example, you want to switch between a set of channels all matched by an ID, for example : `channel.{$channel_id}`, using getListeners(), it will not switch to the other channels, it seems only to care about listening on the first call over the getListeners() method. Perhaps I'm doing something wrong but checking the return value of the getListeners() as the ID changes shows that it does indeed change the array, but the channels it's listening on do not seem to change.

### Exact steps to reproduce
1. Make an event `php artisan make:event ActivityLogged`
2. Set it to broadcastAs `activity-logged` and to broadcast on the desired channel with the ID appended.
3. Make a variable in your livewire component which changes dynamically (modelled by a select, perhaps)
4. Set up the getListeners method to generate the array of listeners with the ID of the desired channel selected
5. Dispatch the event

### Stripped-down, copy-pastable code snippets
(I hope these are concise enough)
Event:
```
<?php

class ActivityLogged implements ShouldBroadcast
{
    use Dispatchable, InteractsWithSockets, SerializesModels;

    public $id;

    public function broadcastAs()
    {
        return 'activity-logged';
    }
  
    public function __construct($id)
    {
        $this->id = $id;
    }

    public function broadcastOn()
    {
        return new PrivateChannel('activity-log.'.$this->id);
    }
}
```
Component:
```
<?php
use Livewire\Component;

class ActivityLog extends Component{

    public $selectedChannelId;

    public function getListeners()
    {
        return isset($this->selectedChannelId) ? [
            "echo-private:activity-log.{$this->selectedChannelId},.activity-logged" => "refresh"
             ] : [];
    }
    public function refresh(){
    }

    public function addActivityLog(){
        event(new ActivityLogged($this->selectedChannelId));
    }
    public function render(){
        return view('livewire.activity-log');
    }
}
```
**Context**
- Livewire version: 2.0 (according to composer.json)
- Laravel version: 8.25.0
- Alpine version: 2.8.0
- Browser: Chrome

**Comment by ArtMin96** [comment]
I think it would be better to use the ``` $refresh ``` method instead of ``` refresh ``` and not create a new method.

**Comment by ArtMin96** [comment]
I think Echo isn't reactive like vue so you're going to have to handle this yourself.
You are going to have to look at the active channels and then add/remove as you make changes.

Something like this.
```
protected function getListeners() 
{
    return [
        "echo-private:activity-log.{$this->selectedChannelId},joining" => 'joining',
        "echo-private:activity-log.{$this->selectedChannelId},leaving" => 'leaving',
    ];
}
```

Take a look at this one, maybe it can help you.
https://stackoverflow.com/questions/63795123/catching-event-in-laravel-with-livewire-and-echo

**Reply by RMK147** [reply]
@francoism90 How did you do that?

**Reply by francoism90** [reply]
@RMK147 I misunderstood the VueJS part. The given solution of @ArtMin96 seems to work, but I haven't used dynamic keys for quite some time now.

**Comment by waterbuckit** [comment]
> I think it would be better to use the `$refresh` method instead of `refresh` and not create a new method.

This is just a demonstration of doing something on an event so doesn't really have bearing on anything but thanks for the tip.

I'd also like to note that `selectedChannelId` is not necessarily set at the time of mounting, as the example shows in the docs. Because I need to jump between channels.

In the end, I've used this as a work around, whereby I leave the channel previously on when a new one is chosen but it's pretty unpleasant to have a script sitting on the page, I'd much rather that livewire's echo listeners could be reactive:

Blade Component: 

```html
<div>
       {{-- some input modelling the selected channel --}}
       
        <script>
        window.addEventListener("activity-log-channel-chosen", (e)=> {
            if(e.detail){
                
                if(e.detail.previous){
                    window.Echo.leave(`activity-log.${e.detail.previous}`);
                }

                window.Echo.private(`activity-log.${e.detail.new}`)
                    .listen('.activity-logged', (e) => {
                        console.log('activity-logged');
                        Livewire.emit('activity-logged');
                    })

             }
        })
       </script>
</div>
```
Livewire Component:
```php
class ActivityLog extends Component {

    public $selectedChannelId;
    public $previousChannelId;

    protected $listeners = [
        "activity-logged" => "refresh",
    ];

    public function updatedSelectedChannelId(){
        $this->previousChannelId = $this->selectedChannelId;
    }

    public function updatingSelectedChannelId($id){
            $this->dispatchBrowserEvent("activity-log-channel-chosen", 
                    ["new" => $id, "previous" => $this->previousChannelId ??  null]);
    }

    public function refresh(){
    }

}
```

**Comment by renepardon** [comment]
I think I have a similar problem. I create an item and then I want to listen to all those item changes/events as well in the same component:

```php
    public function getListeners()
    {
        /** @var User $user */
        $user = auth()->user();

        // This will reload the table after an item X was created/deleted
        $listeners = [
            "echo-private:settings.{$user->id},.XCreated" => 'refetch',
            "echo-private:settings.{$user->id},.XDeleted" => 'refetch',
        ];

        // These channels should be joined as well as soon as there is a new settingsX id
        foreach ($user->settingsX->pluck('id') as $xId) {
            $listeners["echo-private:provisioning.{$xId},.XReadyForProvisioning"] = 'refetch';
            $listeners["echo-private:provisioning.{$xId},.ProvisioningCreated"] = 'refetch';
            $listeners["echo-private:provisioning.{$xId},.XCreated"] = 'refetch';
        }

        return $listeners;
    }
```

But looks like **getListeners()** is only called on inital render/mount. Is there a way to trigger getListeners again to update the listeners also in the frontend?

**Reply by 94Joe** [reply]
I'm currently facing the same problem, did you find a solution to this ?

**Comment by rstanislav** [comment]
Same problem, using Echo - it does not work, at best i'm getting Undefined index error from livewire. 

What is the point of dynamic listeners if they are not working ?

**Comment by mjjabarullah** [comment]
hello all, i am  making web socket chat using laravel livewire and laravel wesockets. i am facing same issue , the issue is when i am switching the chat room dynamically livewire get listeners changes but message event not get listened. but when i reload the whole window then listeners work as indented. any solution for this.

this is my coponen get listener
protected function getListeners()
    {
        return [
            'user-joined' => 'userJoined',
            'hide-loader' => 'hideLoader',
            "echo:chat.$this->roomId,MessageEvent" => 'messageReceived'
        ];
    }

i am joining the channel using this

window.Echo.join(`chat.${roomId}`)
        .here(users => {   
        })

it is working at the time of page started.. but when i chnage the channel the event not getting fired.

**Reply by rtaylor82** [reply]
Any answer here? My livewire component needs to have listener changed based on chat room selected.  Calling the mount() function doesn't update the listeners...

**Reply by mjjabarullah** [reply]
You cannot do like so. Livewire listeners are not reactive. You should manually do some logic

**Reply by mjjabarullah** [reply]
In your java script you should have like this and listen listeners in livewire window.Echo.join(`chat.${roomId}`)
    .subscribed(()=>{
        window.livewire.emit('subscribed')
    })
    .here(users => {
        window.livewire.emit('here', users)
    })
    .joining(user => {
        window.livewire.emit('joining', user)
        // console.log("joining" + JSON.stringify(user))
    })
    .leaving(user => {
        window.livewire.emit('leaving', user)
        // console.log("joining" + JSON.stringify(user))
    })
    .listen('NewMessage', message => {
        window.livewire.emit('newMessage', message)
        // console.log("newMessage" + JSON.stringify(message))
    })
    .listen('MessageDelete', data => {
        window.livewire.emit('removeMessage', data)
        // console.log("deleteMessage" + JSON.stringify(data))
    })
    .error(error => {
        sweetAlert(error)
    })

**Comment by abhiburk** [comment]
I am facing similar kinda of problem. Can getListeners be refreshed or updated ?

**Comment by RMK147** [comment]
Has anyone figured a way how to do this yet without having to add Echo in the JS?

**Comment by prykris** [comment]
Same issue here, so I must manually add JS to the component and register listeners.

**Reply by francoism90** [reply]
Could you give an example on what you're trying to do?

**Reply by prykris** [reply]
I have fixed it.

```blade
@script
<script>
    const initialUUID = @json($article?->uuid);

    if (initialUUID) {
        setupArticleChannel(initialUUID);
    }

    // Handle article creation
    $wire.on('generation.free-article.created', (e) => {
        const generationId = e[0].uuid;

   
        const url = new URL(window.location.href);
        url.searchParams.set('article', generationId);
        window.history.pushState({path: url.href}, '', url.href);

        setupArticleChannel(generationId);
    });

    function setupArticleChannel(uuid) {
        // Cleanup existing channel if any
        if (window.articleChannel) {
            window.Echo.leaveChannel(window.articleChannel.name);
        }

        window.articleChannel = Echo.channel(`generations.freeArticle.${uuid}`)
            .listen('GenerationProgressedEvent', (event) => {
                console.log('GenerationProgressedEvent', event);
                $wire.generationProgressed(event);
            })
            .listen('GenerationExceptionEvent', (event) => {
                console.log('GenerationExceptionEvent', event);
                $wire.generationException(event);
            });
    }
</script>
@endscript
```

Works fine.

**Reply by prykris** [reply]
It's just weird that Livewire doesn't update the listener as the state changes as if listeners can't.

**Reply by francoism90** [reply]
But what do you mean by state change?

```php
// routes/web.php (plain example)

/post/{post}
/article/{article}
```

If the page (route) changes, the listener should change as well. Or is this not happening?

If you listen per item (components), this may not be the best strategie. This will create multiple listeners, and possible increase server load/connection limits.

If you want to listen for user/auth events, it's better to listen on the user channel, and dispatch a refresh or update on the Livewire component. You can check the event in your Livewire component (e.g. `subscription-started`, `favorite`, `saved` and such).

You can still listen for article/post updates on a page, those are the two channels you need.

**Reply by prykris** [reply]
My component creates and assigns the model to its properties, which I called the state change because that's what is happening. Without me having Javascript on the component to see this happening and in turn subscribe to the channel specifically for that model nothing will change unless I refresh the page and only then the listeners I return from `Component::getListeners()` are registered. So it seems that these listeners are registered once and only once at the moment the component gets mounted.

To circumvent this, an event is being dispatched to the front-end and is manually subscribing to the channel; I do this in both cases: on event and page load based on the component's exposed model property value.

I am not switching routes if that is what you are referring to. All I needed was to save the model UUID in the URL to reference a model in the database which will get updated by a Job in Laravel queue. It had to act more as a SPA in my case. I see no reason to reload the whole page just to replace a single element in a DOM.

**Reply by francoism90** [reply]
> [..] on event and page load based on the component's exposed model property value.

This is not going to work unfortunately. You'll be creating multiple listeners for multiple items (I assume it's a list of multiple posts/models?).

Livewire isn't reloading the whole page, that's against the idea of it's design. When you dispatch `$refresh`, it will replace the updated content, not the full page. It does this by using the `wire:key` to keep track of the components that are sent over the wire. It may replace other parts, but you can use `wire:ignore` for creating exceptions.

I don't know your design or solution, but you could add a second Livewire-component, that simple is loaded by a key, and replaced in the DOM on state changes:

```blade
<div id="parent">
    <livewire:post-article :key="$uuid" />
</div>
```

The uuid can be stored in the parent (mostly controller), and updated when needed. The listener is defined in the `post-article`, which should be destroyed when you refresh the `$uuid`, which will make it dynamic.

Please note Livewire does support SPA like features already. I use this to switch between posts, and you can use `@persists` if you want to keep something active between those route changes.

**Reply by prykris** [reply]
I appreciate your help and concerns. I have a single component of this kind on the page at a time. It is not an issue, but I did notice some funky behavior when I tried to register a listener per component when I had like 10 per page. So I'll keep that in mind and come back to your answer for more help when I get to that part of the application.

Edit: your suggestion of `wire:ignore` did come in handy since I had a WYSIWYG editor that was initialized inside the component, but it's DOM kept getting destroyed by the re-rendering.

**Comment by theticraft** [comment]
Now it's 2025. Any answer in the pipeline?

**Reply by francoism90** [reply]
@theticraft What exactly is your issue?

**Reply by theticraft** [reply]
I have a chat that dynamically changes the room.

But if I choose a room, it doesn't subscribe to the new Channel. However, if I implement it like this
```<livewire:chat.chat :chat="\App\Models\Chat::withTrashed()->where('name', 'klsfdkghkdfjl')->first()" />```
It works just fine. So I think it can't subscribe to the channel

For some reference, if needed:

new message event

```
namespace App\Events\chats;

use App\Models\Chat;
use Illuminate\Broadcasting\Channel;
use Illuminate\Broadcasting\InteractsWithSockets;
use Illuminate\Contracts\Broadcasting\ShouldBroadcastNow;
use Illuminate\Foundation\Events\Dispatchable;
use Illuminate\Queue\SerializesModels;

class new_message implements ShouldBroadcastNow
{
    use Dispatchable, InteractsWithSockets, SerializesModels;

    public Chat $chat;

    public function __construct(Chat $chat)
    {
        $this->chat = $chat;
    }

    public function broadcastOn()
    {
        return new Channel("chat_{$this->chat->id}");
    }
}
```

livewire chooser (choose the room)

```
namespace App\Livewire\Chat;

use App\Models\Chat;
use Illuminate\Database\Eloquent\Collection;
use Illuminate\Support\Facades\Auth;
use Livewire\Attributes\Js;
use Livewire\Attributes\Locked;
use Livewire\Attributes\On;
use Livewire\Component;

class Chats extends Component
{
    #[Locked]
    public Collection $chats;
    public Chat $chat;

    public function mount() {
        $this->getChats();
    }

    #[On('echo:chat,chats\\new_chat')]
    #[On('echo:chat,chats\\chat_delete')]
    public function getChats()
    {
        $this->chats = Auth::user()->chats()->withTrashed()->get();
    }

    public function click(int $chat)
    {
        $this->dispatch('chat_click', chat: $chat);
    }

    public function placeholder()
    {
        return "<x-loading />";
    }

    public function render()
    {
        return view('livewire.chat.chats');
    }
}
```

livewire chat room

```
namespace App\Livewire\Chat;

use App\Models\Chat as ModelChat;
use Illuminate\Database\Eloquent\Collection;
use Illuminate\Support\Facades\Auth;
use Livewire\Attributes\Locked;
use Livewire\Attributes\Modelable;
use Livewire\Attributes\On;
use Livewire\Component;
use Nette\Utils\Random;

class Chat extends Component
{
    #[Locked]
    public ModelChat $chat;
    #[Locked]
    public Collection $messages;

    public function mount(ModelChat|int|null $chat=null)
    {
        $this->update($chat);
    }
    
    #[On('chat_click')]
    public function update(ModelChat|int|null $chat=null)
    {
        if(gettype($chat) == 'integer') $chat = ModelChat::find($chat);
        if($chat == null || $chat->users()->find(Auth::user()) === null)
        {
            $this->dispatch('chat_cleared');
            $this->render();
            return;
        }
        $this->chat = $chat;
        $this->retrieveMessages();
        $this->dispatch('$refresh');
    }

    public function getListeners()
    {
        $return = [];
        if(!(($this->chat ?? null) === null))  $return["echo:chat_{chat.id},chats\\new_message"] = 'retrieveMessages';
        return $return;
    }

    public function retrieveMessages()
    {
        $this->messages = $this->chat->messages;
    }

    public function placeholder()
    {
        return "<x-loading />";
    }

    #[On('echo:chat,chats\\new_chat')]
    #[On('echo:chat,chats\\chat_delete')]
    public function render()
    {
        return view('livewire.chat.chat');
    }
}
```

**Reply by francoism90** [reply]
I do this with a `key`, basically:

```php
// Livewire/chats/rooms.php

class Rooms extends Component
{
	// You can also use a model here
	public string $room;
}

// Livewire/chats/messages.php

class Messages extends Component
{
	// I actually recommend to use a model here
	public string $room;

	#[Computed(persist: true, seconds: 3600)]
	public function messages(): Collection
	{
		// ..
	}
}
```

In your `rooms.blade.php` for example:

```blade
// build a select with all rooms
// <select wire:model="room">

<livewire:chats.messages :$room :key="$room" />
```

You need to use the `key` to make it refresh on the fly.

I did once submitted a PR with dynamic computed support, but it unfortunately got rejected.
This would make it more dynamic. But now you need to use the key.

**Reply by theticraft** [reply]
I implement the components like this. How do I get the key there?
```<livewire:chat.chats lazy />``` is the chooser
```<livewire:chat.chat />``` is the room itself
```
<div class="col-span-2"><livewire:chat.chats lazy /></div>
<div class="col-span-8"><livewire:chat.chat /></div>
```

So the data gets there and all the messages are loaded, but it doesn't subscribe to the echo channel

**Reply by francoism90** [reply]
The `key` value can be everything, it can even be a random string. But I actually recommend using the model-id (or even better: route-key). You can also use `crc32` (or any hash method) as a Livewire key, this is very useful when your model attributes change as well.

In the `Messages` component you define the listeners `public function getListeners()` method.

You can read more about it over here https://livewire.laravel.com/docs/nesting#forcing-a-child-component-to-re-render, even when it doesn't apply to nesting, it should work in examples like this.

**Reply by theticraft** [reply]
Now I get 404s, because I edited a but more, but thank you for your very fast replies! I appreciate it very much!

**Reply by francoism90** [reply]
Take a good look at the given examples in the docs, and do it step-for-step. :)

The variable should exists, this means it should get it's value from something in the component. You cannot use `mt_rand()` or something.

Do you mean listeners giving a 404 or the component loading?

**Reply by theticraft** [reply]
Yeah, the route defining it self is already wrong. I need to fix it, because it has an optional route parameter, but it isn't needed because of livewires `#[Url]` modifier.
I will write to you tomorrow if needed. 
So thanks for your support ❤️❤️

**Reply by uzair404** [reply]
Any  Updates On this one? it seems like everyone ended up using JavaScript?

**Comment by azzazkhan** [comment]
We're at the end of 2025 and still hitting the wall with this limitation. 💀
